src/main/kotlin/org/rust/ide/annotator/RustHighlightingAnnotator.kt

```diff
@@ -12,7 +12,11 @@ import org.rust.lang.core.psi.util.elementType
 
 // Highlighting logic here should be kept in sync with tags in RustColorSettingsPage
 class RustHighlightingAnnotator : Annotator {
-    override fun annotate(element: PsiElement, holder: AnnotationHolder) = element.accept(object : RustElementVisitor() {
+
+    override fun annotate(element: PsiElement, _holder: AnnotationHolder) {
+        element.accept(object : RustElementVisitor() {
+        var holder = wrap(_holder)
```

Comment:
Ah, I see know the `resolveColor` trick. Hm, I wonder if this can be simplified by using a more side effect free style...

What about something like this:

```Kotlin
data class HighlightingInfo (
    val textAttributesKey: TextAttributesKey
) {
    fun apply(element: PsiElement, holder: AnnotationHolder) {
        holder.createInfoAnnotation(element, null).textAttributes = textAttributesKey
    }
}

override fun annotate(element: PsiElement, holder: AnnotationHolder) {
    if (element is RustReference) {
        val resolve = element.reference.resolve() ?: return
        val highlightingInfo = calculateHighlightingInfo(resolve)
        highlightingInfo.apply(element, holder) 
    }
}

```

?